lms/djangoapps/instructor_task/tasks_helper/grades.py

```diff
@@ -548,6 +552,90 @@ def _graded_scorable_blocks_to_header(cls, course):
 
 
 class ProblemResponses(object):
+
+    @classmethod
+    def _build_problem_list(cls, course_blocks, root, path=None):
+        """
+        Generate a tuple of display names, block location paths and block keys
+        for all problem blocks under the ``root`` block.
+
+        Arguments:
+            course_blocks (BlockStructureBlockData): Block structure for a course.
+            root (UsageKey): This block and its children will be used to generate
+                the problem list
+            path (List[str]): The list of display names for the parent of root block
+
+        Yields:
+            Tuple[str, List[str], UsageKey]: tuple of a block's display name, path, and
+                usage key
+        """
+        display_name = course_blocks.get_xblock_field(root, 'display_name')
+        if path is None:
+            path = [display_name]
+
+        yield display_name, path, root
+
+        for block in course_blocks.get_children(root):
+            display_name = course_blocks.get_xblock_field(block, 'display_name')
+            for result in cls._build_problem_list(course_blocks, block, path + [display_name]):
+                yield result
+
+    @classmethod
+    def _build_student_data(cls, user_id, course_id, problem_location):
+        """
+        Generate a list of problem responses for all problem under the
+        ``problem_location`` root.
+
+        Arguments:
+            user_id (int): The user id for the user generating the report
+            course_id (CourseKey): The ``CourseKey`` for the course whose report
+                is being generated
+            problem_location (str): The generated report will include this
+                block and it child blocks.
+
+        Returns:
+              List[Dict]: Returns a list of dictionaries containing the student
+                data which will be included in the final csv.
+        """
+        problem_key = UsageKey.from_string(problem_location).map_into_course(course_id)
+        user = get_user_model().objects.get(pk=user_id)
+        course_blocks = get_course_blocks(user, problem_key)
+
+        student_data = []
+        max_count = settings.FEATURES.get('MAX_PROBLEM_RESPONSES_COUNT')
+
+        for title, path, block_key in cls._build_problem_list(course_blocks, problem_key):
+            # Chapter and sequential blocks are filtered out since they include state
+            # which isn't useful for this report.
+            if block_key.block_type in ('sequential', 'chapter'):
+                continue
+
+            block = course_blocks.get_xblock(block_key)
+            generate_report_data = None
+
+            if hasattr(block, 'generate_report_data'):
+                generate_report_data = block.generate_report_data
+            elif hasattr(block, 'module_class') and hasattr(block.module_class, 'generate_report_data'):
+                generate_report_data = block.module_class.generate_report_data
```

Comment:
@xitij2000 I don't think `generate_report_data` should be a staticmethod, because then it can't access the XBlock's `Scope.settings` and `Scope.content` fields.

How about using an approach like:

```python
from xmodule.modulestore.django import modulestore
from courseware.user_state_client import DjangoXBlockUserStateClient
from opaque_keys.edx.keys import UsageKey
user_state_client = DjangoXBlockUserStateClient()

block = modulestore().get_item(block_key)
user_state_iterator = user_state_client.iter_all_for_block(block_key)
block.generate_report_data(user_state_iterator)
```

You can try this with both an XBlock and an XModule using this example code I whipped up:
* Drag and Drop: https://github.com/open-craft/xblock-drag-and-drop-v2/commit/953c30c31983b085a3ea0110807e4a7a43e42dea
* Capa: https://github.com/open-craft/edx-platform/commit/a8136e2cf85f65b06927adc597ed378e5cbdbd04

Reply:
@bradenmacdonald  Ah! I think if the method is on the descriptor then there is no problem. My issue was that I only had access to the descriptor or the module class, and I didn't want to code the differnt initialization paths for differnt kinds of blocks in the report generatioon code, so I recommended the static approach.

Reply:
@xitij2000 Yep. Descriptors are not specific to a student, while modules are, so it makes sense for this code to be on a Descriptor only. Newer XBlocks (like drag and drop) aren't split into descriptor vs. module, so for pure XBlocks it doesn't matter. But for older XModules like capa, this sort of code should be on the descriptor if we must choose.